### PR TITLE
Admin user

### PR DESCRIPTION
**Pull request by shernanm**
## Summary

* Endpoint is implemented that allows users to create administrator or update their profile.

* Their respective tests are implemented.

## Trello Card

https://trello.com/c/AmHXviQf

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -42,3 +42,17 @@ exports.getUsersList = (req, res, next) => {
     .then(response => res.send(response))
     .catch(next);
 };
+
+exports.createAdminUser = (req, res, next) => {
+  const user = req.body;
+  logger.info(`createAdminUser method start, request methods: ${req.method}, endpoint: ${req.path},
+  user: ${user.firstName} ${user.lastName}`);
+  req.body.role = 'admin';
```
Comment (raparicio6) [inline app/controllers/user.js]: why this is harcoded?

File: app/controllers/user.js [hunk 2]
```diff
@@ -42,3 +42,17 @@ exports.getUsersList = (req, res, next) => {
...
+  return changeRole(req.body.email)
+    .then(updated =>
```
Comment (raparicio6) [inline app/controllers/user.js]: why does this return an array?

File: app/controllers/user.js [hunk 3]
```diff
@@ -42,3 +42,17 @@ exports.getUsersList = (req, res, next) => {
...
+      updated[0]
+        ? res.status(201).send({ message: `User ${updated[1][0].firstName} updated to admin` })
```
Comment (raparicio6) [inline app/controllers/user.js]: put an unique data to identify the user, such as the email

File: app/services/users.js [hunk 4]
```diff
@@ -40,8 +40,21 @@ exports.findAllUsers = (limit, offset) =>
...
-    attributes: ['id', 'firstName', 'lastName', 'email']
+    attributes: ['id', 'firstName', 'lastName', 'email', 'role']
   }).catch(() => {
     logger.info('Error trying to find the users');
     throw errors.databaseError('Error processing request in database.');
   });
+
+exports.changeRole = email =>
+  User.update(
+    { role: 'admin' },
```
Comment (raparicio6) [inline app/services/users.js]: why this method always change the role to admin? doesnt have to allow to change to any role?

File: test/adminUser.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
+const request = require('supertest');
```
Comment (raparicio6) [inline test/adminUser.test.js]: dont create a new file. make the tests in the user test file in a new describe

**Review by raparicio6 — CHANGES_REQUESTED** [review]

File: app/controllers/user.js [hunk 6]
```diff
@@ -42,3 +43,17 @@ exports.getUsersList = (req, res, next) => {
...
     .catch(next);
 };
+
+exports.createAdminUser = (req, res, next) => {
+  const user = req.body;
+  logger.info(`createAdminUser method start, request methods: ${req.method}, endpoint: ${req.path},
+  user: ${user.firstName} ${user.lastName}`);
+  req.body.role = ADMIN_ROLE;
+  return changeRole(req.body.email)
+    .then(([AffectedColumns]) =>
```
Comment (raparicio6) [inline app/controllers/user.js]: why in this case? use camel case

File: app/middlewares/sessionValidator.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+const { decodedToken } = require('../utils');
+const errors = require('../errors');
+
+exports.isAdminUser = (req, res, next) => {
+  const { role } = decodedToken(req.headers.token);
+  return role === 'admin'
```
Comment (raparicio6) [inline app/middlewares/sessionValidator.js]: use constant

**Review by raparicio6 — CHANGES_REQUESTED** [review]

File: app/services/users.js [hunk 8]
```diff
@@ -40,8 +44,19 @@ exports.findAllUsers = (limit, offset) =>
...
   }).catch(() => {
     logger.info('Error trying to find the users');
     throw errors.databaseError('Error processing request in database.');
   });
+
+exports.changeRole = email =>
+  User.update(
+    { role: ADMIN_ROLE },
+    {
+      where: { email, role: REGULAR_ROLE }
```
Comment (molinajulian) [inline app/services/users.js]: The condition role is unnecessary, email  is  unique

File: migrations/migrations/20190625222111-admin-user-role.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+module.exports = {
+  up: (queryInterface, Sequelize) =>
+    queryInterface.addColumn('users', 'role', {
+      type: Sequelize.STRING,
```
Comment (molinajulian) [inline migrations/migrations/20190625222111-admin-user-role.js]: Check type `ENUM`

File: seeders/20190626154842-first-user-admin.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+'use strict';
```
Comment (molinajulian) [inline seeders/20190626154842-first-user-admin.js]: Delete this migration

File: app/models/user.js [hunk 11]
```diff
@@ -21,6 +21,11 @@ module.exports = (sequelize, DataTypes) => {
       password: {
         allowNull: false,
         type: DataTypes.STRING
+      },
+      role: {
+        allowNull: false,
+        type: DataTypes.STRING,
```
Comment (molinajulian) [inline app/models/user.js]: Change for `ENUM`

**Review by molinajulian — CHANGES_REQUESTED** [review]

File: app/services/users.js (hunk 8 above)
Comment (shernanm) [reply app/services/users.js]: I do this so that it does not always update

File: app/services/users.js [hunk 12]
```diff
@@ -2,20 +2,35 @@ const { User } = require('../models');
...
+      }
+      return passwordEncryption(user);
+    })
+    .then(response =>
+      User.create(response).catch(() => {
+        throw errors.databaseError('Error processing request in database.');
+      })
+    )
+    .catch(error => {
+      logger.info(`Error trying to create the user ${user.firstName} ${user.lastName}`);
```
Comment (raparicio6) [inline app/services/users.js]: logger.error

File: app/services/users.js [hunk 13]
```diff
@@ -2,20 +2,35 @@ const { User } = require('../models');
...
+  exports
+    .findUserBy({ email: user.email })
+    .then(foundUser => {
+      if (foundUser) {
+        throw errors.badRequest('The email is already registered.');
+      }
+      return passwordEncryption(user);
+    })
+    .then(response =>
+      User.create(response).catch(() => {
```
Comment (raparicio6) [inline app/services/users.js]: this not sounds good. a response seems a http response and you dont give a http response to the database to create an user. put the names clearer please

**Review by molinajulian — APPROVED** [review]

File: app/services/users.js [hunk 14]
```diff
@@ -40,8 +55,19 @@ exports.findAllUsers = (limit, offset) =>
...
-    attributes: ['id', 'firstName', 'lastName', 'email']
+    attributes: ['id', 'firstName', 'lastName', 'email', 'role']
   }).catch(() => {
     logger.info('Error trying to find the users');
     throw errors.databaseError('Error processing request in database.');
   });
+
+exports.changeRole = email =>
+  User.update(
+    { role: ADMIN_ROLE },
```
Comment (raparicio6) [inline app/services/users.js]: did you see my comment of making this method available to change to any role?
here you are making the change to admin role always, its not generic

File: app/services/users.js [hunk 15]
```diff
@@ -40,8 +55,19 @@ exports.findAllUsers = (limit, offset) =>
...
+    {
+      where: { email, role: REGULAR_ROLE }
+    }
+  ).catch(() => {
+    logger.info('Error trying to update the user');
```
Comment (raparicio6) [inline app/services/users.js]: logger.error

**Review by raparicio6 — CHANGES_REQUESTED** [review]

**Review by raparicio6 — APPROVED** [review]

**Pull request merged**